### PR TITLE
win, fs: mkdir really return UV_EINVAL for invalid names

### PR DESCRIPTION
**Pull request by seishun**
Makes uv_fs_mkdir return UV_EINVAL for invalid directory names instead
of UV_ENOENT.

Ref: https://github.com/nodejs/node/issues/31177

I'm not happy about locally overriding `uv_translate_sys_error` but I can't think of a better approach.

### CODE REVIEW

**Comment by seishun** [comment]
@bzoz

**Comment by vtjnash** [comment]
FWIW, this was already changed on master (https://github.com/libuv/libuv/pull/1088) a couple years ago, and is a known_issue in nodejs (https://github.com/nodejs/node/pull/28569) but seems like @saghul said it would break compatibility to change on v1.x series?

**Comment by seishun** [comment]
That's for ERROR_INVALID_NAME, this is for ERROR_DIRECTORY.

**Comment by vtjnash** [comment]
Ah, right—I read that backwards. So maybe on master/v2 we want to alter the mapping (like #1088) and on v1.x we do this (like #2375)?

**Comment by seishun** [comment]
Mapping ERROR_DIRECTORY to UV_EINVAL would be consistent with `_mkdir` and `_rmdir`. Unfortunately CreateProcess sets ERROR_DIRECTORY if `lpCurrentDirectory` is set to a non-existent directory (see 9f07a3673b301da99bd31f2d793983caad6e3428, it seems @bnoordhuis misunderstood the cause of the breakage). I think the proper solution would be to add a case for UV_EINVAL [here](https://github.com/nodejs/node/blob/e65bed1b7e273d1db6ba5905c8f68338b06cd27a/lib/internal/child_process.js#L386-L390).

**Comment by vtjnash** [comment]
Does nodejs still want this? Just needs a rebase in that case.

**Comment by seishun** [comment]
Sorry, I'm not interested in working on this anymore. Feel free to take over.

**Comment by vtjnash** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit/2010/
CI+Nodejs: waiting on https://github.com/nodejs/node/pull/34751

Okay, it looked good, I was just hoping a nodejs dev would chime in a lgtm, since it'll fix their issue.

File: test/test-fs.c [hunk 1]
```diff
@@ -4376,6 +4376,8 @@ TEST_IMPL(fs_invalid_mkdir_name) {
   loop = uv_default_loop();
   r = uv_fs_mkdir(loop, &req, "invalid>", 0, NULL);
   ASSERT(r == UV_EINVAL);
+  r = uv_fs_mkdir(loop, &req, "test:lol", 0, NULL);
+  ASSERT(r == UV_EINVAL);
```
Comment (bnoordhuis) [inline test/test-fs.c:4380]:
```suggestion
  ASSERT_EQ(UV_EINVAL, uv_fs_mkdir(loop, &req, "test:lol", 0, NULL));
```

**Review by bnoordhuis — APPROVED** [review]
LGTM and thanks, Nikolai!

**Comment by Xstoudi** [comment]
I can confirm this patch fix the mentionned issues on Node.js.

**Review by santigimeno — APPROVED** [review]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2101/

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/dd8662b6d28236b60bb419390ea083695bbb2173. Thanks and really sorry for the long delay.

**Pull request closed without merging**